### ISSUE TITLE
kokkos-kernels: compilation error "redefinition of struct SVD"

### ISSUE THREAD

**Issue by maartenarnst** [trilinos/Trilinos#12891]
@trilinos/kokkos-kernels
@lucbv 
@ndellingwood 

### Description

We are seeing compilation errors in kokkos-kernels, in a Trilinos build with both Lapack ad MKL enabled, following the latest snapshot.  

The compilation errors look like:

```
In file included from /Trilinos-sources/packages/kokkos-kernels/lapack/impl/KokkosLapack_svd_spec.hpp:154,
                 from /Trilinos-sources/build-with-GNU-OpenMP-amd64/packages/kokkos-kernels/lapack/eti/generated_specializations_cpp/svd/Lapack_svd_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_OPENMP_MEMSPACE_HOSTSPACE.cpp:20:
 /Trilinos-sources/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp:273:10: error: redefinition of 'struct KokkosLapack::Impl::SVD<Kokkos::Serial, Kokkos::View<float**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >, Kokkos::View<float*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >, Kokkos::View<float**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >, Kokkos::View<float**, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<1> >, true, false>'
```

```
/Trilinos-sources/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp:273:10: note: in definition of macro 'KOKKOSLAPACK_SVD_MKL'
|   struct SVD<
```

```
/Trilinos-sources/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp:174:1: note: in expansion of macro 'KOKKOSLAPACK_SVD_LAPACK'
| KOKKOSLAPACK_SVD_LAPACK(float, Kokkos::LayoutLeft, Kokkos::Serial)
```

It appears that the issue is that the recent snapshot introduces two structs `SVD` with the same name and template parameters, protected by preprocessor conditionals. And when both Lapack and MKL are provided as TPLs, both definitions become active, and one becomes a redefinition of the other.

- https://github.com/trilinos/Trilinos/blob/864738b09cfcd361267c7205287daaa2a25063e1/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp#L112-L117

- https://github.com/trilinos/Trilinos/blob/864738b09cfcd361267c7205287daaa2a25063e1/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp#L271-L276

**Comment by ndellingwood** [comment]
@maartenarnst sorry for the inconvenience, I'm going to work on a reproducing but can you share details of your configuration as well?
We had tested kokkos-kernels release-candidate with Trilinos using intel/2023.2.0 with mkl for our blas+lapack libs, with blas and lapack tpls enabled but I did not run into this issue.
I suspect the way I configure to use mkl may differ from your setup, for example I do the following:

```
export BLAS_LIBRARIES="${MKLROOT}/lib/intel64/libmkl_intel_lp64.a;${MKLROOT}/lib/intel64/libmkl_intel_thread.a;${MKLROOT}/lib/intel64/libmkl_core.a"
export LAPACK_LIBRARIES=${BLAS_LIBRARIES}

...
# begin cmake line
...
  -DTPL_ENABLE_BLAS:BOOL=ON \
  -DTPL_BLAS_LIBRARIES:PATH="${BLAS_LIBRARIES}" \
  -DTPL_ENABLE_LAPACK:BOOL=ON \
  -DTPL_LAPACK_LIBRARIES:PATH="${LAPACK_LIBRARIES}" \
...
```

**Comment by ndellingwood** [comment]
Reproduced the error, will look into resolving

**Comment by ndellingwood** [comment]
I reproduced by explicitly enabling MKL (`-DTPL_ENABLE_MKL=ON`) and pointing to the include and lib dirs, along with the BLAS and LAPACK fragments I posted above.

I tested the change below, which allowed me to compile the kokkos-kernels and unit tests pass without error (still need to compile and test more of Trilinos). My intent was to disambiguate between the LAPACK and MKL `SVD` specializations by deferring to the MKL specialization when both TPLs are enabled (by setting the `tpl_spec_avail` template parameter as false in the LAPACK specialization when MKL is enabled):

```diff
diff --git a/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp b/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
index bc23068c57f..b0b963d6f50 100644
--- a/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
+++ b/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
@@ -109,6 +109,14 @@ void lapackSvdWrapper(const ExecutionSpace& /* space */, const char jobu[],
   }
 }
 
+constexpr bool svd_lapack_enable() {
+#ifdef KOKKOSKERNELS_ENABLE_TPL_MKL
+  return false;
+#else
+  return true;
+#endif
+}
+
 #define KOKKOSLAPACK_SVD_LAPACK(SCALAR, LAYOUT, EXEC_SPACE)                 \
   template <>                                                               \
   struct SVD<                                                               \
@@ -125,7 +133,7 @@ void lapackSvdWrapper(const ExecutionSpace& /* space */, const char jobu[],
       Kokkos::View<SCALAR**, LAYOUT,                                        \
                    Kokkos::Device<EXEC_SPACE, Kokkos::HostSpace>,           \
                    Kokkos::MemoryTraits<Kokkos::Unmanaged>>,                \
-      true,                                                                 \
+      svd_lapack_enable(),                                                  \
       svd_eti_spec_avail<                                                   \
           EXEC_SPACE,                                                       \
           Kokkos::View<SCALAR**, LAYOUT,                                    \
```

I'll check with other kokkos-kernels devs on how best to proceed

**Comment by ndellingwood** [comment]
Much simpler change for now, thanks for suggestion by @brian-kelley 

```diff
diff --git a/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp b/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
index bc23068c57f..7f8ccca5bac 100644
--- a/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
+++ b/packages/kokkos-kernels/lapack/tpls/KokkosLapack_svd_tpl_spec_decl.hpp
@@ -41,7 +41,7 @@ inline void svd_print_specialization() {
 }  // namespace KokkosLapack
 
 // LAPACK
-#ifdef KOKKOSKERNELS_ENABLE_TPL_LAPACK
+#if defined(KOKKOSKERNELS_ENABLE_TPL_LAPACK) && !defined(KOKKOSKERNELS_ENABLE_TPL_MKL)
 #include "KokkosLapack_Host_tpl.hpp"
 
 namespace KokkosLapack {
```

Submitting PRs to kokkos-kernels, after those are confirmed to pass testing we'll get snapshot into Trilinos

**Comment by ndellingwood** [comment]
@maartenarnst please let us know if this is resolved for you by #12892 (just merged); I was able to reproduce the issue locally and confirmed that PR fixed it with my configuration

**Comment by maartenarnst** [comment]
Hi @ndellingwood and @brian-kelley. Thanks a lot for solving this. In the mean time, I've run our builds the latest develop, and, yes, indeed, your PR solves the issue in all our builds.
